pinot-core/src/main/java/com/linkedin/pinot/core/segment/store/SegmentLocalFSDirectory.java

```diff
@@ -215,16 +223,62 @@ protected File starTreeIndexFile() {
 
   private PinotDataBuffer getIndexForColumn(String column, ColumnIndexType type)
       throws IOException {
+    PinotDataBuffer buffer;
     switch (type) {
       case DICTIONARY:
-        return columnIndexDirectory.getDictionaryBufferFor(column);
+        buffer = columnIndexDirectory.getDictionaryBufferFor(column);
+        break;
       case FORWARD_INDEX:
-        return columnIndexDirectory.getForwardIndexBufferFor(column);
+        buffer = columnIndexDirectory.getForwardIndexBufferFor(column);
+        break;
       case INVERTED_INDEX:
-        return columnIndexDirectory.getInvertedIndexBufferFor(column);
+        buffer = columnIndexDirectory.getInvertedIndexBufferFor(column);
+        break;
       default:
         throw new RuntimeException("Unknown index type: " + type.name());
     }
+    if (readMode == ReadMode.mmap) {
+      prefetchMmapData(buffer);
+    }
+    return buffer;
+  }
+
+  private void prefetchMmapData(PinotDataBuffer buffer) {
+    // mmap mode causes high number of major page faults after server restart.
+    // This impacts latency especially for prod "online" use cases that require low latency.
+    // This function proactively loads pages in memory to lower the variance in
+    // latencies after server startup.
+
+    // This has to handle two different data size profiles
+    // 1. Servers with data size close to main memory size
+    // 2. Servers with very large data sizes (terabytes)
+    // To prevent it from loading terabytes of data on startup, we put a limit
+    // on the number of pages this will prefetch (OS will do something more on top of this)
+    // The logic here is as follows:
+    // Server doesn't know total data size it is expected to serve. So this will
+    // load all data till 2/3rd (PREFETCH_SLOWDOWN_PCT) of the configured limit. After that it will only
+    // read the header page. We read headers because that has more frequently accessed
+    // information which will have bigger impact on the latency. This can go over the limit
+    // because it doesn't stop at any point. But that's not an issue considering this is
+    // an optimization.
+
+    // Prefetch limit and slowdown percentage are arbitrary
+    if (prefetchedPages.get() >= MAX_MMAP_PREFETCH_PAGES) {
+      return;
+    }
+
+    final long prefetchSlowdownPageLimit = (long) (PREFETCH_SLOWDOWN_PCT * MAX_MMAP_PREFETCH_PAGES);
```

Comment:
it can be but it's better to derive it rather than multiple interdependent globals. This allows us to change the logic more easily in localized manner.